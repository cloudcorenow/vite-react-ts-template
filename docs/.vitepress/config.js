// @ts-check

/**
 * @type {import('vitepress').UserConfig}
 */
module.exports = {
  title: 'Vite⚡',
  description: 'Next Generation Frontend Tooling',
  head: [
    [
      'style',
      {},
      'img { border-radius: 10px }' + 'h1.title { margin-left: 0.5em }'
    ]
  ],
  themeConfig: {
    repo: 'vitejs/vite',
    docsDir: 'docs',
    docsBranch: 'main',
    editLinks: true,
    editLinkText: 'Suggest changes to this page',

    locales: {
      '/': {
        lang: 'en-US',
        label: 'English',
        selectText: 'Languages',
        nav: [
          { text: 'Guide & APIs', link: '/guide/' },
          { text: 'Config Reference', link: '/config/' }
        ]
      },
      '/zh/': {
        lang: 'zh-CN',
        label: '中文',
        selectText: '选择语言',
        editLinkText: '为此页提供修改建议',
        nav: [
          { text: '指引 & API', link: '/zh/guide/' },
          { text: '配置参考', link: '/zh/config/' }
        ]
      }
    },

    sidebar: {
      '/config/': 'auto',
      '/zh/config/': 'auto',
      '/zh/': [
        {
          text: '指引',
          children: [
            {
              text: '介绍',
              link: '/zh/guide/introduction'
            },
            {
              text: '开始',
              link: '/zh/guide/'
            },
            {
              text: '功能',
              link: '/zh/guide/features'
            },
            {
              text: '依赖预构建',
              link: '/zh/guide/dep-pre-bundling'
            },
            {
              text: '构建生产版本',
              link: '/zh/guide/build'
            },
            {
              text: '环境变量与模式',
              link: '/zh/guide/env-and-mode'
            },
            {
              text: '后端集成',
              link: '/zh/guide/backend-integration'
            },
            {
              text: '比较',
              link: '/zh/guide/comparisons'
            },
            {
              text: '从 v1 迁移',
              link: '/zh/guide/migration'
            }
          ]
        },
        {
          text: 'API',
          children: [
            {
              text: '配置参考',
              link: '/zh/config/'
            },
            {
              text: '插件 API',
              link: '/zh/guide/api-plugin'
            },
            {
              text: 'HMR API',
              link: '/zh/guide/api-hmr'
            },
            {
              text: 'JavaScript API',
              link: '/zh/guide/api-javascript'
            }
          ]
        }
      ],
      // catch-all fallback
      '/': [
        {
          text: 'Guide',
          children: [
            {
              text: 'Introduction',
              link: '/guide/introduction'
            },
            {
              text: 'Getting Started',
              link: '/guide/'
            },
            {
              text: 'Features',
              link: '/guide/features'
            },
            {
              text: 'Dependency Pre-Bundling',
              link: '/guide/dep-pre-bundling'
            },
            {
              text: 'Building for Production',
              link: '/guide/build'
            },
            {
              text: 'Env Variables and Modes',
              link: '/guide/env-and-mode'
            },
            {
              text: 'Backend Integration',
              link: '/guide/backend-integration'
            },
            {
              text: 'Comparisons',
              link: '/guide/comparisons'
            },
            {
              text: 'Migration from v1',
              link: '/guide/migration'
            }
          ]
        },
        {
          text: 'APIs',
          children: [
            {
              text: 'Config Reference',
              link: '/config/'
            },
            {
              text: 'Plugin API',
              link: '/guide/api-plugin'
            },
            {
              text: 'HMR API',
              link: '/guide/api-hmr'
            },
            {
              text: 'JavaScript API',
              link: '/guide/api-javascript'
            }
          ]
        }
      ]
    }
  },
  locales: {
    '/': {
      lang: 'en-US'
    },
    '/zh': {
      lang: 'zh-CN',
      description: '下一代前端开发与构建工具'
    }
  }
}
