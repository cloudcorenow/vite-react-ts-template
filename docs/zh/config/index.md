# 配置 Vite

## 配置文件

### 配置文件解析

当以命令行方式运行 `vite` 时，Vite 会自动解析 [项目根目录](/zh/guide/#project-root) 下名为 `vite.config.js` 的文件。

最基础的配置文件是这样的：

```js
// vite.config.js
export default {
  // 配置选项
}
```

注意，即使项目没有在 `package.json` 中开启 `type: "module"` ，Vite 也支持在配置文件中使用 ESM 语法。在这种情况下，配置文件是在加载之前自动预处理的。

你可以显式地通过 `--config` 命令行选项指定一个配置文件（解析会相对于 `cwd` 路径）

```bash
vite --config my-config.js
```

### 配置智能提示

因为 Vite 本身附带 Typescript 类型，所以你可以通过 IDE 和 jsdoc 的配合来进行智能提示：

```js
/**
 * type {import('vite').UserConfig}
 */
export default {
  // ...
}
```

Vite 同样支持 TS 配置文件。你可以直接使用 `vite.config.ts`：

```ts
import { defineConfig } from 'vite'

export default defineConfig({
  // ...
})
```

### 场景配置

如果配置文件需要基于命令（`serve` 或 `build`）或者不同场景与 [模式](/zh/guide/env-and-mode) 来决定选项，可以选择导出这样一个函数：

```js
export default ({ command, mode }) => {
  if (command === 'serve') {
    return {
      // serve 独有配置
    }
  } else {
    return {
      // build 独有配置
    }
  }
}
```

## 共享配置

### alias

- **类型：**
  `Record<string, string> | Array<{ find: string | RegExp, replacement: string }>`

  将会被传递到 `@rollup/plugin-alias` 作为它的 [entries](https://github.com/rollup/plugins/tree/master/packages/alias#entries)。也可以是一个对象，或一个 `{ find, replacement }` 的数组.

  更高级的自定义解析方法可以通过 [插件](/zh/guide/api-plugin) 实现。

### define

- **类型：** `Record<string, string>`

  定义全局变量替换。在开发期间，entries 将被定义为全局变量，在构建期间被静态替换。

### root

- **类型：** `string`
- **默认：** `process.cwd()`

  项目根目录。可以是一个绝对路径，或者一个相对于该配置文件本身的路径。

  更多细节请见 [项目根目录](/zh/guide/#项目根目录)。

### mode

- **Type:** `string`
- **Default:** `'development'` for serve, `'production'` for build

  在配置中特别指定将会替换掉 serve 和 build 时的默认模式。还可以通过命令行选项 `--mode` 覆写该值。

  更多细节请见 [环境变量与模式](/zh/guide/env-and-mode)。

### plugins

- **类型：** ` (Plugin | Plugin[])[]`

  要使用的插件数组。

  获取关于 Vite 插件的更多细节请见 [插件 API](/zh/guide/api-plugin) 。

### css.modules

- **类型：**

  ```ts
  interface CSSModulesOptions {
    scopeBehaviour?: 'global' | 'local'
    globalModulePaths?: string[]
    generateScopedName?:
      | string
      | ((name: string, filename: string, css: string) => string)
    hashPrefix?: string
    /**
     * 默认：'camelCaseOnly'
     */
    localsConvention?: 'camelCase' | 'camelCaseOnly' | 'dashes' | 'dashesOnly'
  }
  ```

  配置 CSS modules 的行为。选项将被传递给 [postcss-modules](https://github.com/css-modules/postcss-modules)。

### css.preprocessorOptions

- **类型：** `Record<string, object>`

  指定传递给 CSS 预处理器的选项。例如:

  ```js
  export default {
    css: {
      preprocessorOptions: {
        scss: {
          additionalData: `$injectedColor: orange;`
        }
      }
    }
  }
  ```

### esbuild

- **类型：** `ESBuildOptions | false`

  `ESBuildOptions` 继承自 [esbuild 转换选项](https://esbuild.github.io/api/#transform-api)。最常见的用例是自定义 JSX：

  ```js
  export default {
    esbuild: {
      jsxFactory: 'h',
      jsxFragment: 'Fragment'
    }
  }
  ```

  默认情况下，ESbuild 应用在 `ts`、`jsx`、`tsx` 文件。你可以通过 `esbuild.include` 和 `esbuild.exclude` 对其进行配置，它们两个配置的类型是`string | RegExp | (string | RegExp)[]`。

  设置成 `false` 可以禁用 ESbuild 转换（默认应用于 `.ts`. `.tsx` 和 `.jsx` 文件）。

  此外，你还可以通过`esbuild.jsxInject`来自动为每一个被 ESbuild 转换的文件注入 JSX helper。

  ```js
  export default {
    esbuild: {
      jsxInject: `import React from 'react'`
    }
  }
  ```

### assetsInclude

- **类型：** `string | RegExp | (string | RegExp)[]`
- **相关内容：** [Asset Handling](/zh/guide/features#asset-handling)

  指定其他文件类型作为静态资源处理（这样导入它们就会返回解析后的 URL）

### transformInclude

- **类型：** `string | RegExp | (string | RegExp)[]`

  默认情况下，代码中所有静态可分析的 `import` 请求都会被静态地视为转换管道的一部分。当然，如果你使用动态导入来导入非 js 类型的文件，例如:

  ```js
  // dynamicPath 是一个非 JS 文件类型，例如 "./foo.gql"
  import(dynamicPath).then(/* ... */)
  ```

  Vite 无法知道文件需要被转换为 JavaScript（还是直接被视为静态文件提供服务）。`transfromInclude` 配置项允许你显式地声明文件类型，让它始终被转换或者当成 JavaScript 进行服务。

### dedupe

- **类型：** `string[]`

  如果你在你的应用程序中有相同依赖的副本（比如 monorepos），使用这个选项来强制 Vite 总是将列出的依赖关系解析到相同的副本（从项目根目录)。

### logLevel

- **类型：** `'info' | 'warn' | 'error' | 'silent'`

  调整控制台输出的级别，默认为 `'info'`.

## Server Options

### server.host

- **类型：** `string`

  指定服务器主机名

### server.port

- **类型：** `number`

  指定服务器端口。注意：如果端口已经被使用，Vite 会自动尝试下一个可用的端口，所以这可能不是服务器最终监听的实际端口。

### server.https

- **类型：** `boolean | https.ServerOptions`

  启用 TLS + HTTP/2。注意当 [`server.proxy` option](#server-proxy) 也被使用时，将会仅使用 TLS。

  这个值也可以是一个传递给 `https.createServer()` 的 [选项对象](https://nodejs.org/api/https.html#https_https_createserver_options_requestlistener)。

### server.open

- **类型：** `boolean`

  在服务器启动时自动在浏览器中打开应用程序。

### server.proxy

- **类型：** `Record<string, string | ProxyOptions>`

  为开发服务器配置自定义代理规则。期望接收一个 `{ key: options }` 对象。使用 [`http-proxy`](https://github.com/http-party/node-http-proxy)。完整选项详见 [此处](https://github.com/http-party/node-http-proxy#options).

  **示例：**

  ```js
  export default {
    server: {
      proxy: {
        // 字符串简写写法
        '/foo': 'http://localhost:4567/foo',
        // 选项写法
        '/api': {
          target: 'http://jsonplaceholder.typicode.com',
          changeOrigin: true,
          rewrite: (path) => path.replace(/^\/api/, '')
        }
      }
    }
  }
  ```

### server.cors

- **类型：** `boolean | CorsOptions`

  为开发服务器配置 CORS。默认启用并允许任何源，传递一个 [选项对象](https://github.com/expressjs/cors) 来调整行为或设为 `false` 表示禁用。

### server.force

- **类型：** `boolean`
- **相关内容：** [Dependency Pre-Bundling](/zh/guide/dep-pre-bundling)

  设置为 `true` 强制使依赖预构建。

### server.hmr

- **类型：** `boolean | { protocol?: string, host?: string, port?: number, path?: string, timeout?: number, overlay?: boolean }`

  禁用或配置 HMR 连接（用于 HMR websocket 必须使用不同的 http 服务器地址的情况）。

  设置 `server.hmr.overlay` 为 `false` 可以禁用服务器错误遮罩层。

### server.watch

- **类型：** `object`

  传递给 [chokidar](https://github.com/paulmillr/chokidar#api) 的文件系统监视器选项。

## Build Options

### build.base

- **类型：** `string`
- **默认：** `/`

  在生产环境中的 public base 路径。注意，路径应该以 `/` 开头和结尾。查看 [public base 路径](/zh/guide/build#public-base-path) 来获取更多细节。

### build.target

- **类型：** `string`
- **默认：** `es2020`
- **相关内容：:** [浏览器兼容性](/zh/guide/build#browser-compatibility)

  设置构建后浏览器兼容的 ES 版本。转换过程将会由 esbuild 执行，最低目标支持 `es2015`。目标也可以是带有浏览器的版本，例如 `chrome58` 或 `safari11`，或由多个目标组成的一个数组。

  注意，如果代码包含不能被 `esbuild` 安全地编译的特性，那么构建将会失败。查看 [esbuild 文档](https://esbuild.github.io/api/#target) 获取更多细节。

### build.outDir

- **类型：** `string`
- **默认：** `dist`

  指定输出路径（相对于 [项目根目录](/zh/guide/#项目根目录)).

### build.assetsDir

- **类型：** `string`
- **默认：** `assets`

  指定生成静态资源的存放路径（相对于 `build.outDir`）。

### build.assetsInlineLimit

- **类型：** `number`
- **默认：** `4096` (4kb)

  小于此阈值的导入或引用资源将内联为 base64 编码，以避免额外的 http 请求。设置为 `0` 可以完全禁用此项。

### build.cssCodeSplit

- **类型：** `boolean`
- **默认：** `true`

  启用/禁用 CSS 代码拆分。当启用时，在异步 chunk 中导入的 CSS 将内联到异步 chunk 本身，并在块加载时插入。

  如果禁用，整个项目中的所有 CSS 将被提取到一个 CSS 文件中。

### build.sourcemap

- **类型：** `boolean`
- **默认：** `false`

  构建后是否生成 source map 文件。

### build.rollupOptions

- **类型：** [`RollupOptions`](https://rollupjs.org/guide/en/#big-list-of-options)

  自定义底层的 Rollup 打包配置。这与从 Rollup 配置文件导出的选项相同，并将与 Vite 的内部 Rollup 选项合并。查看 [Rollup 选项文档](https://rollupjs.org/guide/en/#big-list-of-options) 获取更多细节。

### build.lib

- **类型：** `{ entry: string, name?: string, formats?: ('es' | 'cjs' | 'umd' | 'iife')[] }`
- **相关内容：** [Library Mode](/zh/guide/build#库模式)

  构建为库。`entry` 是必须的因为库不可以使用 HTML 作为入口。`name` 则是暴露的全局变量，并且在 `formats` 包含 `'umd'` 或 `'iife'` 时是必须的。默认 `formats` 是 `['es', 'umd']`。

### build.manifest

- **类型：** `boolean`
- **默认：** `false`
- **相关内容：** [后端集成](/zh/guide/backend-integration)

  当设置为 `true`，构建后将会生成 `manifest.json` 文件，映射没有被 hash 的资源文件名和它们的 hash 版本。可以为一些服务器框架渲染时提供正确的资源引入链接。

### build.minify

- **类型：** `boolean | 'terser' | 'esbuild'`
- **默认：** `'terser'`

  设置为 `false` 可以禁用最小化混淆，或是用来指定使用哪种混淆器。默认为 [Terser](https://github.com/terser/terser)，虽然 Terser 相对较慢，但大多数情况下构建后的文件体积更小。ESbuild 最小化混淆更快但构建后的文件相对更大。

### build.terserOptions

- **类型：** `TerserOptions`

  传递给 Terser 的更多 [minify 选项](https://terser.org/docs/api-reference#minify-options)。

### build.write

- **类型：** `boolean`
- **默认：** `true`

  设置为 `false` 来禁用将构建后的文件写入磁盘。这常用于 [编程式地调用 `build()`](/zh/guide/api-javascript#build) 在写入磁盘之前，需要对构建后的文件进行进一步处理。

## 依赖优化选项

- **相关内容：** [Dependency Pre-Bundling](/zh/guide/dep-pre-bundling)

### optimizeDeps.include

- **类型：** `string[]`

  在预构建中强制包含的依赖项。

### optimizeDeps.exclude

- **类型：** `string[]`

  在预构建中强制排除的依赖项。

### optimizeDeps.link

- **类型：** `string[]`

  在预构建中显式地将依赖项视为链接源。注意 Vite 2.0 会自动检测 “链接包”（解析后路径不在 `node_modules` 内依赖的)，所以只有在极少数情况下才需要这样做。

### optimizeDeps.allowNodeBuiltins

- **类型：** `string[]`

  一个导入节点内置组件的依赖项列表，但在浏览器中并不实际使用它们。会抑制相关的警告。

  A list of dependencies that imports Node built-ins, but do not actually use them in browsers. Suppresses related warnings.

### optimizeDeps.auto

- **类型：** `boolean`
- **默认：** `true`

  服务器启动时自动运行依赖预构建。若设置为 `false` 则禁用。
