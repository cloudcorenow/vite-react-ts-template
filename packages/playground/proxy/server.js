const http = require('http')

async function createServer(root, isProd) {
  http
    .createServer((req, res) => {
      res.end()
    })
    .listen(8081, () => {
      console.log('http://localhos:8081/')
    })
  http
    .createServer((req, res) => {
      res.writeHead(200, { 'Content-Type': 'application/json' })
      res.end(JSON.stringify({ key: 'router' }))
    })
    .listen(8080, () => {
      console.log('http://localhos:8080/')
    })
}

exports.createServer = createServer
