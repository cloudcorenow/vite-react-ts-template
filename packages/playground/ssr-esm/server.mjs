import path from 'path'
import { fileURLToPath } from 'url'

import { install } from 'source-map-support'
install()

const __dirname = path.dirname(fileURLToPath(import.meta.url))

const vite = await import('vite')
const server = await vite.createServer({
  root: path.join(__dirname, 'app'),
  // mode: 'production',
  ssr: {
    external: ['cjs-package', 'esm-package']
  },
  server: {
    middlewareMode: 'ssr'
  }
})

const entryModule = await server.ssrLoadModule('/entry-server.jsx')
console.log(entryModule.default)
