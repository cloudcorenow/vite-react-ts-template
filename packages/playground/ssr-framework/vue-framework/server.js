const express = require('express')
const { createPageRender } = require('./createPageRender.js')
const { getViteConfig } = require('./getViteConfig.js')
const vite = require('vite')

module.exports.startServer = startServer

async function startServer(root, port, isProduction) {
  const app = express()

  let viteServer
  if (isProduction) {
    app.use(express.static(`${root}/dist/client`, { index: false }))
  } else {
    viteServer = await vite.createServer({
      ...getViteConfig(root),
      logLevel: 'warn',
      optimizeDeps: { entries: ['**/pages/*.vue'] },
      server: { middlewareMode: true }
    })
    app.use(viteServer.middlewares)
  }

  const renderPage = createPageRender(viteServer, root, isProduction)
  app.get('*', async (req, res, next) => {
    const url = req.originalUrl
    const html = await renderPage(url)
    if (html === null) return next()
    res.send(html)
  })

  return new Promise((resolve) => {
    const server = app.listen(port, () => {
      const stopServer = async () => {
        server.close()
        if (viteServer) await viteServer.close()
      }
      resolve(stopServer)
    })
  })
}
