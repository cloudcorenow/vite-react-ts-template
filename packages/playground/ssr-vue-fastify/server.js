const fs = require('fs')
const { resolve } = require('path')
const Fastify = require('fastify')
const FastifyVitePlugin = require('./plugin')

const isTest = process.env.NODE_ENV === 'test' || !!process.env.VITE_TEST_BUILD

async function createServer(
  root = process.cwd(),
  isProd = process.env.NODE_ENV === 'production'
) {
  const fastify = new Fastify()

  await fastify.register(FastifyVitePlugin, {
    dev: true,
    rootDir: resolve(__dirname),
    srcDir: resolve(__dirname, 'src'),
  })

  fastify.vite.get('/with-data', {
    ssrData (req) {
      return { message: `Hello from ${req.raw.url}` }
    }
  })
  fastify.get('/*', fastify.vite.handler)

  return { fastify, vite: fastify.$vite }
}

if (!isTest) {
  createServer().then(({ fastify }) =>
    fastify.listen(3000, (err, address) => {
      if (err) {
        console.error(err)
        process.exit(1)
      }
      console.log(`Server listening on ${address}`)
    })
  )
}

// for test use
exports.createServer = createServer
