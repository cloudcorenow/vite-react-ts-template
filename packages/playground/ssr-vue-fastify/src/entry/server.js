import { createApp } from '../main'
import { renderToString } from '@vue/server-renderer'
import devalue from '@nuxt/devalue'

export async function render (req, url, manifest, ssrDataKey) {
  const { ctx, app, router } = createApp(req)
  
  app.config.globalProperties[ssrDataKey] = req[ssrDataKey]

  router.push(url)

  await router.isReady()

  const html = await renderToString(app, ctx)
  const preloadLinks = renderPreloadLinks(ctx.modules, manifest)

  return [html, preloadLinks]
}

function renderPreloadLinks (modules, manifest) {
  let links = ''
  const seen = new Set()
  modules.forEach((id) => {
    const files = manifest[id]
    if (files) {
      files.forEach((file) => {
        if (!seen.has(file)) {
          seen.add(file)
          links += renderPreloadLink(file)
        }
      })
    }
  })
  return links
}

function renderPreloadLink(file) {
  if (file.endsWith('.js')) {
    return `<link rel="modulepreload" crossorigin href="${file}">`
  } else if (file.endsWith('.css')) {
    return `<link rel="stylesheet" href="${file}">`
  } else {
    // TODO
    return ''
  }
}
