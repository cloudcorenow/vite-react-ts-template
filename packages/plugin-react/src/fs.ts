import fs from 'fs'

export function readJsonFile(file: string, throwError?: boolean): any {
  let content: string
  try {
    content = fs.readFileSync(file, 'utf8')
  } catch (e) {
    if (throwError) throw e
    return
  }
  return JSON.parse(content)
}
