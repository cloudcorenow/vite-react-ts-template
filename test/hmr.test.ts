import { traverseEsModules, urlResolver } from 'es-module-traversal'
import { once } from 'events'
import execa from 'execa'
import fs from 'fs-extra'
import path from 'path'
import { URL } from 'url'
import WebSocket from 'ws'
import { UserConfig } from '../dist/node/config'
import { createServer } from '../dist/node/server'
import playgroundConfig from '../playground/vite.config'

const tempDir = path.resolve('temp')
const fixtureDir = path.resolve('playground')
const PORT = 4000

jest.setTimeout(100000)

const config: UserConfig = {
  ...playgroundConfig,
  alias: {
    alias: '/alias/aliased',
    '/@alias/': path.resolve(tempDir, 'alias/aliased-dir')
  },
  port: PORT,
  root: tempDir
}

type TestCase = {
  name: string
  path: string
  expectedMessagesCount?: number
  replacer: (content: string) => string
}

const testCases: TestCase[] = [
  {
    name: 'simple vue file',
    path: 'hmr/TestHmr.vue',
    expectedMessagesCount: 1,
    replacer: (content) => {
      return content.replace('</script>', '\n</script>')
    }
  },
  {
    name: 'vue js propagation',
    path: 'hmr/testHmrPropagation.js',
    expectedMessagesCount: 1,
    replacer: (content) => {
      return content.replace('</script>', '\n</script>')
    }
  },
  {
    name: 'vue dynamic import propagation',
    path: 'hmr/testHmrPropagationDynamicImport.js',
    expectedMessagesCount: 1,
    replacer: (content) => {
      return content + '\n\n'
    }
  },
  {
    name: 'vue manual hmr',
    path: 'hmr/testHmrManual.js',
    expectedMessagesCount: 1,
    replacer: (content) => {
      return content + '\n\n'
    }
  },
  {
    name: 'vue alias hmr',
    path: 'alias/aliased/index.js',
    expectedMessagesCount: 1,
    replacer: (content) => {
      return content + '\n\n'
    }
  },
  {
    name: 'vue css import',
    path: 'css-@import/testCssAtImportFromScript.css',
    expectedMessagesCount: 1,
    replacer: (content) => {
      return content + '\n\n'
    }
  },
  {
    name: 'vue css modules import',
    path: 'css-@import/testCssModuleAtImportFromScript.module.css',
    expectedMessagesCount: 2,
    replacer: (content) => {
      return content + '\n\n'
    }
  },
  {
    name: 'vue css style @import hmr',
    path: 'css-@import/testCssAtImportFromStyle.css',
    expectedMessagesCount: 1,
    replacer: (content) => {
      return content + '\n\n'
    }
  },
  {
    name: 'vue css style @import modules hmr',
    path: 'css-@import/testCssModuleAtImportFromStyle.module.css',
    expectedMessagesCount: 1,
    replacer: (content) => {
      return content + '\n\n'
    }
  },
  {
    name: 'css @import',
    path: 'css-@import/imported.css',
    expectedMessagesCount: 1,
    replacer: (content) => {
      return content + '\n\n'
    }
  }
]

beforeAll(async () => {
  try {
    await fs.remove(tempDir)
  } catch (e) {}
  await fs.ensureDir(tempDir)
  await fs.copy(fixtureDir, tempDir, {
    filter: (file) => !/dist|node_modules/.test(file)
  })
  await execa('yarn', { cwd: tempDir, stdio: 'inherit' })
  await execa('yarn', {
    cwd: path.join(tempDir, 'optimize-linked'),
    stdio: 'inherit'
  })
})

afterAll(async () => {
  try {
    await fs.remove(tempDir)
  } catch (e) {}
})

describe('playground hmr', () => {
  const baseUrl = `http://localhost:${PORT}`

  const root = tempDir

  let server
  beforeAll(async () => {
    server = createServer(config)
    server.listen(PORT)
    await once(server, 'listening')
    // console.log('sleeping')
    // await sleep(1000000 * 1000)
  })
  afterAll(() => {
    server.close()
  })

  for (let testCase of testCases) {
    test(testCase.name, async () => {
      const traversedFiles = await traverseEsModules({
        entryPoints: [new URL('/main.js', baseUrl).toString()],
        resolver: urlResolver({
          root,
          baseUrl
        })
      })
      // console.log(traversedFiles.map((x) => x.importPath))
      const messages = await getWsMessages({
        doing: async () => {
          await updateFile(path.resolve(root, testCase.path), testCase.replacer)
        },
        expectedMessagesCount: testCase.expectedMessagesCount
      })
      expect(messages.map(normalizeHmrMessage)).toMatchSnapshot()
    })
  }
})

async function updateFile(compPath, replacer) {
  try {
    const content = await fs.readFile(compPath, 'utf-8')
    await fs.writeFile(compPath, replacer(content))
  } catch (e) {
    throw new Error(`could not update ${compPath}: ${e}`)
  }
}

async function getWsMessages({
  doing,
  expectedMessagesCount = Infinity,
  timeout = 500,
  port = PORT
}) {
  const ws = new WebSocket(`http://localhost:${port}`, 'vite-hmr')
  await once(ws, 'open')
  await doing()
  const messages = []
  ws.addEventListener('message', ({ data }) => {
    const payload = JSON.parse(data)
    if (payload.type === 'connected') return
    if (payload.type === 'multi') {
      return messages.push(...payload.updates)
    }
    return messages.push(payload)
  })
  await Promise.race([
    waitUntil(() => messages.length === expectedMessagesCount),
    sleep(timeout)
  ])
  ws.close()
  await once(ws, 'close')
  return messages
}

const sleep = (n) => new Promise((r) => setTimeout(r, n))

async function waitUntil(check) {
  while (!check()) {
    await sleep(50)
  }
}

const normalizeHmrMessage = (message) => {
  const ignoreKeys = ['timestamp']
  const validKeys = Object.keys(message).filter((k) => !ignoreKeys.includes(k))
  return Object.assign({}, ...validKeys.map((k) => ({ [k]: message[k] })))
}
