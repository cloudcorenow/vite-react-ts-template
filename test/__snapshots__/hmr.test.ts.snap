// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`playground hmr css @import 1`] = `
Array [
  Object {
    "changeSrcPath": "/css-@import/testCssAtImportFromScript.css",
    "path": "/css-@import/testCssAtImportFromScript.css",
    "type": "style-update",
  },
]
`;

exports[`playground hmr simple vue file 1`] = `
Array [
  Object {
    "changeSrcPath": "/hmr/TestHmr.vue",
    "path": "/hmr/TestHmr.vue",
    "type": "vue-reload",
  },
]
`;

exports[`playground hmr vue alias hmr 1`] = `
Array [
  Object {
    "changeSrcPath": "/alias/aliased/index.js",
    "path": "/alias/TestAlias.vue",
    "type": "vue-reload",
  },
]
`;

exports[`playground hmr vue css import 1`] = `
Array [
  Object {
    "changeSrcPath": "/css-@import/testCssAtImportFromScript.css",
    "path": "/css-@import/testCssAtImportFromScript.css",
    "type": "style-update",
  },
]
`;

exports[`playground hmr vue css modules import 1`] = `
Array [
  Object {
    "changeSrcPath": "/css-@import/testCssModuleAtImportFromScript.module.css",
    "path": "/css-@import/TestCssAtImport.vue",
    "type": "vue-reload",
  },
  Object {
    "changeSrcPath": "/css-@import/testCssModuleAtImportFromScript.module.css",
    "path": "/css-@import/testCssModuleAtImportFromScript.module.css",
    "type": "style-update",
  },
]
`;

exports[`playground hmr vue css style @import hmr 1`] = `
Array [
  Object {
    "changeSrcPath": "/css-@import/TestCssAtImport.vue?type=style&index=0",
    "path": "/css-@import/TestCssAtImport.vue?type=style&index=0",
    "type": "style-update",
  },
]
`;

exports[`playground hmr vue css style @import modules hmr 1`] = `
Array [
  Object {
    "changeSrcPath": "/css-@import/TestCssAtImport.vue?type=style&index=1",
    "path": "/css-@import/TestCssAtImport.vue?type=style&index=1",
    "type": "style-update",
  },
]
`;

exports[`playground hmr vue dynamic import propagation 1`] = `
Array [
  Object {
    "changeSrcPath": "/hmr/testHmrPropagationDynamicImport.js",
    "path": "/hmr/TestHmr.vue",
    "type": "vue-reload",
  },
]
`;

exports[`playground hmr vue js propagation 1`] = `
Array [
  Object {
    "changeSrcPath": "/hmr/testHmrPropagation.js",
    "path": "/hmr/TestHmr.vue",
    "type": "vue-reload",
  },
]
`;

exports[`playground hmr vue manual hmr 1`] = `
Array [
  Object {
    "changeSrcPath": "/hmr/testHmrManual.js",
    "path": "/hmr/testHmrManual.js",
    "type": "js-update",
  },
]
`;
